### DISCUSSION TITLE
Why is Bevy's Coordinate System Right-Handed  Y-Up?

### DISCUSSION THREAD

**Discussion by cart** [bevyengine/bevy#10488] [Q&A] [answered]
Why is Bevy's coordinate system right-handed y-up?

**Comment by cart** [comment, marked as answer]
This can be broken up into two questions, but the optimal solve requires considering both:

* Why is Bevy right handed?
* Why is Bevy y-up?

This really only comes down to these choices (from the perspective of our current right handed coordinate system):

1. Do what is natural for right handed 2D / 2.5D and make Y-Up the global assumption across 2D and 3D, at the cost of making 3D _slightly_ weird by making  -Z "forward" semantically. This is a hidden implementation detail for most people.
2. Do what is maximally natural for right handed 3D and make Z-up the global assumption, at the cost of making 2D _very very_ weird (2D gamedevs think in z-levels for depth, not y-levels. Y is always the vertical axis, not Z)
3. Fracture Up, Down, Forward, and Back concepts in Bevy into `UP_2D` (Y) and `UP_3D` (Z). These would no longer be universal in Bevy.  This increases our weirdness budget/api surface and makes documenting/answering questions hard ("User: what is up in Bevy?", followed by "Helpful Bevian: Thats a complicated question ... [dumps a bunch of info]").  It also puts 2.5D / platformer style games in an awkward position. Would those devs pick `UP_3D` because they have 3D models? Or would they pick `UP_2D` because they want to use the conventions natural to their genre? 
4. Switch handedness, which breaks compatibility with math conventions and most of the ecosystem that we care about (ex: all 3D modeling programs, godot)

IMO (1) is the optimal solve (the current approach) by a wide margin

Unity and unreal dodged this problem by doing (4). But they created way more painful "ecosystem compatibility" problems as a result.

Freya's "gamedev ecosystem coordinate system" chart helps illustrate why right-handed is the right call from an ecosystem perspective:
![image](https://github.com/bevyengine/bevy/assets/2694663/3f2d1284-95f8-4f90-a8c9-64c83be5bcb8)

**Reply by thmxv** [reply]
This does not make much sense. 

Pretty much every math book ever written use:
- X axis going left to right
- Y axis going near to far
- Z axis going up
- Right hand rule

Using this would  make perfect sense for 3D and for 2D. In 2D, the viewing direction (eyes to page) is the top down direction (-Z)

It also would be a better compatibility with the ecosystem:
- All math books use that
- All engineering/CAD software use that
- All architecture software use that

Also this explanation of "why use unnatural conventions"  as "because it is natural" does not hold much.

The only way this "Y up" is natural is if you are confused by "bottom of the page" (near) and "top of the page" (far) and think this means "bottom" and "top". Or if you read your math books in bed (holding the book vertically not horizontally).

Anyway great engine none the less. And rust: great language even if if uses "Vec" as a type for an array which have nothing to do with linear algebra.

Unrelated but in the same vibe: What is wrong with the hardware (IEEE 754) where `1.0/0.0 == Inf` and `1.0/-0.0 == -Inf` where obviously both should be `NaN`?

**Reply by emkkla** [reply]
It depends on your point of view.
In CGI, VFX, or animation, the X-axis represents left/right, and the Y-axis represents down/up, similar to 2D mathematical graphs where X is the horizontal axis, Y is the vertical axis, and Z adds depth (positive values toward the front, negative values toward the rear).
For many of us, having Z as the up axis feels weird.
I guess, the ideal solution would be an agnostic engine that lets us choose among the four coordinate systems.

**Reply by Clonkex** [reply]
For me, it feels most natural for X to be left-right, Y to be up-down (as on any computer screen), and Z to add depth to that screen. And then it feels most intuitive for positive movement on each axis to represent the positive-feeling direction in 3D space. For instance, "right" feels more positive than "left" (as a right-handed person), "up" feels more positive than "down", and "forward" feels more positive than "backward". Thus I find Bevy's choice of -Z == forward to be very confusing to reason about (especially when considering cameras - do they just point towards -Z, so moving them in the positive direction moves them backwards?).

**Reply by TransparentWhit** [reply]
> Anyway great engine none the less. And rust: great language even if if uses "Vec" as a type for an array which have nothing to do with linear algebra.

If I recall correctly, the vec thing was a historical issue originally introduced by cpp (for reasons i can not comprehend)

**Comment by rlidwka** [comment]
I feel the need to answer this, since working with coordinate systems was the second most annoying thing in Bevy I encountered last project (*the most annoying by far was delayed transform propagation*).

I actually had 4 different coordinate systems in one project. Bevy as game engine, blender as asset editor (from which I can't simply copy displayed XYZ coordinates to bevy). *And then, since it was a robotics project, there're two standards, NED and ENU, because americans can't possibly agree with europeans on anything related to standards, and I had to use both to interop with different things.*

The most natural system (for me at least, but hopefully many agree) is: X forward, Y right, Z up. Is that what Unreal does? If so, I can very much respect them picking usability choice over legacy support.

X forward, Y right, Z up is great, very intuitive, with a small downside. It's left handed. Is it actually a downside? What does chirality change actually? You need to flip cross-products (including normals), and that's about it.

If you don't want to do that, maybe consider compatibility with 3d modeling tools. X forward, Y **left**, Z up (or Y forward, X right, Z up) is going to be at least compatible with Blender, which has XY as ground plane. It is a bit annoying in yaw/pitch/roll calculations, but at least it removes coordinate transforms from blender imports (that's probably the most widely used 3d editor nowadays).

I seriously can't understand Y being vertical. For 3D it is natural to have XY as ground plane. For 2D it is natural to have XY as screen plane. Only difference here is camera rotation: for 2D it looks towards -Z, and for 3D first-person shooter it would look towards X.

And finally, the coordinate system I'm considering to use nowadays is: X forward, Y right, Z **down**. This corresponds to a standard in robotics called NED (North-East-Down). It is right handed, it has natural XY. Z is inverted to preserve chirality, but that's easy to get used to.

**Reply by rlidwka** [reply]
So here's what I'm suggesting:

1. switch to XY being a ground plane in 3D with Z vertical
2. since point 1 is probably not gonna happen, provide tools to convert between coordinate systems
3. give Unity award for picking *the worst* coordinate system (Y-up, and left-handed), here's a fun forum thread I just dug up: https://forum.unity.com/threads/z-should-be-up.308187/

**Reply by afonsolage** [reply]
You argument is most based on personal preferences. For me, the most natural system is right-handed Y-up (which Bevy already uses), mostly because my background is Minecraft, Godot and a lot of games/engines based on OpenGL.

**Reply by Barugon** [reply]
I agree with @rlidwka. It would also be easier to think about when creating assets in Blender, since everything would use the same coordinate system.

**Reply by SolraBizna** [reply]
At the end of the day, everything you render must be transformed into window space, where X is right and Y is up. No exceptions. Z-up means that which axis is up changes during the rendering process, which is why it is a major peeve of mine. Window space is the only coordinate system that *must* factor into your transformations. What do you gain by fighting it?

As for negative Z being forward, I find it helps to, instead of holding on to the concept of "deepness", think in terms of "foregroundness": something with a higher Z is in the foreground relative to something with a lower Z. Instead of "negative Z forward", remember "Z toward me". (CSS has the same model for Z and it never confused me there.)

Many 3D modeling tools use Z-up because of their origins in CAD. Most computer-controlled work tools hover over the piece "looking" down, and do their work layer by layer. For these tools, X and Y are still left/right and down/up, it's just that *gravity doesn't point down*. For a PDP-7 in 1965, which shipped with only four kibiwords of memory and struggled to achieve 200 000 integer adds per second, it made sense to put the transformation burden on the humans doing the modeling. With each new iteration of CAD software, compatibility with existing CAD workflows was more important than outside conventions for axis definitions. If only they had known this would still be causing trouble sixty years later...

**Reply by rlidwka** [reply]
> At the end of the day, everything you render must be transformed into window space, where X is right and Y is up. No exceptions.

Y is down in [win32](https://learn.microsoft.com/en-us/windows/win32/gdi/window-coordinate-system), as well as in [css](https://developer.mozilla.org/en-US/docs/Web/CSS/CSSOM_view/Coordinate_systems). But this might depend on each particular device and driver.

> What do you gain by fighting it?

More intuitive coordinate system to work with 3d.

What do you gain by **not** fighting it? A few CPU instructions in best case?

> As for negative Z being forward, I find it helps to, instead of holding on to the concept of "deepness", think in terms of "foregroundness": something with a higher Z is in the foreground relative to something with a lower Z

I work with 3D objects. 3D objects have forward, lateral, vertical directions. 3D objects do not have "deepness", "foregroundness", or any of the sort.

I'm okay with **viewport** having XY coordinate system, because viewport is inherently 2-dimensional. In fact, you can't have it in any other way, really. But why do you force this concept onto 3D?

> For a PDP-7 in 1965, which shipped with only four kibiwords of memory

I'm not sure where you copy-pasted this from, but "[kibiword](https://en.wiktionary.org/wiki/kibiword)" is not a thing, the unit is spelled [kiloword](https://en.wiktionary.org/wiki/kiloword).

**Reply by SolraBizna** [reply]
> Y is down in [win32](https://learn.microsoft.com/en-us/windows/win32/gdi/window-coordinate-system), as well as in [css](https://developer.mozilla.org/en-US/docs/Web/CSS/CSSOM_view/Coordinate_systems). But this might depend on each particular device and driver.

Y was down in nearly every 2D-only graphics system. That's still on the up/down axis. [Edit: If I'm being honest, disagreements between Y-down and Y-up cause me a lot more trouble than disagreements between Y-up and Z-up when handedness is the same.]

> I'm okay with viewport having XY coordinate system, because viewport is inherently 2-dimensional. In fact, you can't have it in any other way, really. But why do you force this concept onto 3D?

Because I'm projecting this 3D scene onto a 2D surface.

> I'm not sure where you copy-pasted this from, but "[kibiword](https://en.wiktionary.org/wiki/kibiword)" is not a thing, the unit is spelled [kiloword](https://en.wiktionary.org/wiki/kiloword).

"kibi" is the ~~SI~~ [edit: [IEC](https://en.wikipedia.org/wiki/Binary_prefix)] binary prefix for 1024, analogous to "kilo", the SI prefix for 1000.

**Reply by Barugon** [reply]
For 2D, I personally think of a traditional cell animation setup, where the camera is looking down onto the sheets.

**Reply by rlidwka** [reply]
> Because I'm projecting this 3D scene onto a 2D surface.

You can project something onto XY plane or XZ plane, math works out the same. That's inner details of the renderer anyway.

Why should I as a user care about this detail if I'm describing 3d world?

> "kibi" is the SI binary prefix for 1024, analogous to "kilo", the SI prefix for 1000.

Neither "byte", and especially not "word" are SI/SI-derived units, so SI standards are not applicable here.

*Going on a tangent here: Computer science does not use SI prefixes. Current recommendation, as far as I remember, is to use "kilobyte" as a unit of storage capacity, which is shorthanded as "KB" or occasionally "KiB" for clarity, and **always** means 2^10 bytes. I can provide citations, but I'm afraid it has nothing to do with Bevy really. Edit: I vaguely remember "decimal kilobyte" being a unit, but it is rarely used, so doesn't have a shorthand.*

**Reply by SolraBizna** [reply]
> You can project something onto XY plane or XZ plane, math works out the same. That's inner details of the renderer anyway.

I think I have been unclear. When I say I'm projecting something onto a 2D plane, I'm talking specifically about the rendering step; the only part of a 3D rendering pipeline that is not optional.

If you are in a situation where Z-up makes more sense or is more convenient to you, go right ahead and use Z-up—if you're starting from a Y-up frame, that's just one view rotation away. What bothers me is when an engine (or modelling program) *forces* Z-up, because then it *forces* having two definitions for up instead of one—in cases where Y-up is more convenient, this leads to two rotations instead of zero. There's no significant efficiency cost either way, but it adds a layer to the math, which makes it easier to make mistakes. (Such as with Minecraft, which routinely and to my great displeasure switches handedness randomly during rendering.)

Anyway.

If we want to be really classy, we could be like POV-ray, and [make the user specify right, up, and forward vectors](http://www.povray.org/documentation/3.7.0/r3_4.html#r3_4_2), permitting any handedness and axis definition the user finds convenient. Just as long as we *don't* be like POV-ray in making left-handedness the default.

---

Further discussion of units and prefixes in [SI vs. IEC prefixes and bytes](https://github.com/bevyengine/bevy/discussions/10584), where it is *less* off-topic.

**Reply by BatmanAoD** [reply]
> The most natural system (for me at least, but hopefully many agree) is: X forward, Y right, Z up.

I realize that "intuitive" is almost entirely subjective, but I'd like to know what I'm missing here because neither "x forward" nor "y right" seems in any way obvious or intuitive.
